libopae/src/error.c

```diff
@@ -0,0 +1,286 @@
+			     uint32_t error_num,
+			     struct fpga_error_info *error_info)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	uint32_t i = 0;
+
+	if (!error_info) {
+		FPGA_MSG("error_info is NULL");
+		return FPGA_INVALID_PARAM;
+	}
+
+	// TODO: make thread-safe
+
+	// TODO: should we populate this if user hasn't created a
+	//       properties structure for this token?
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		if (i == error_num) {
+			memcpy_s(error_info, sizeof(struct fpga_error_info), &p->info, sizeof(struct fpga_error_info));
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+/* files and directories to ignore when looking for errors */
+const char *errors_exclude[] = {
+	"revision",
+	"uevent",
+	"power",
+	"clear"
+};
+#define NUM_ERRORS_EXCLUDE (sizeof(errors_exclude) / sizeof(char *))
+
+/* returns the number of error entries added to `list` */
+uint32_t build_error_list(const char *path, struct error_list **list)
+{
+	struct dirent *de;
+	DIR *dir;
+	struct stat st;
+	char basedir[FILENAME_MAX];
+	int len = strlen(path);
+	uint32_t n = 0;
+	unsigned int i;
+	struct error_list **el = list;
+
+	if (len+1 > FILENAME_MAX) {
+		FPGA_MSG("path too long");
+		return 0;
+	}
+
+	strncpy(basedir, path, FILENAME_MAX-1);
+	basedir[len++] = '/';
+
+	dir = opendir(path);
+	if (!dir) {
+		FPGA_MSG("unable to open %s", path);
+		return 0;
+	}
+
+	while ((de = readdir(dir))) {
+		// skip hidden ('.*') files (includes "." and "..")
+		if (de->d_name[0] == '.')
+			continue;
+
+		// skip names on blacklist
+		for (i = 0; i < NUM_ERRORS_EXCLUDE; i++) {
+			if (strcmp(de->d_name, errors_exclude[i]) == 0) {  // FIXME: SAFE?
+				break;
+			}
+		}
+		if (i < NUM_ERRORS_EXCLUDE)
+			continue;
+
+		// build absolute path
+		strncpy(basedir + len, de->d_name, FILENAME_MAX - len);
+
+		// try accessing file/dir
+		if (lstat(basedir, &st) == -1) {
+			FPGA_MSG("can't stat %s", basedir);
+			continue;
+		}
+
+		// skip symlinks
+		if (S_ISLNK(st.st_mode))
+			continue;
+
+		// recursively dive into subdirectories
+		if (S_ISDIR(st.st_mode)) {
+			n += build_error_list(basedir, el);
+			continue;
+		}
+
+		// not blacklisted, not hidden, accessible, no symlink, no dir -> count and append it!
+		n++;
+		if (!el)	// no list
+			continue;
+
+		// append error info to list
+		struct error_list *new_entry = malloc(sizeof(struct error_list));
+		if (!new_entry) {
+			FPGA_MSG("can't allocate memory");
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+		new_entry->next = NULL;
+		// see if error can be cleared (currently only errors called "errors" corresponding
+		// to a "clear" file can)
+		new_entry->info.can_clear = false;
+		if (strcmp(de->d_name, "errors") == 0) {    // FIXME: SAFE
```

Comment:
Will this `FIXME` be addressed in another pull request?

Reply:
No - I'll address that in this PR; it'll get shot down by static analysis otherwise.